### PR TITLE
Improve logs in primary storage removal process

### PR DESCRIPTION
**Pull request by hsato03**
### Description

While listing volumes from a primary storage, volumes from system VMs are not listed, making the primary storage removal process difficult as an exception is thrown if the storage still has active volumes when trying to be deleted.

For this reason, logs have been added to the primary storage removal process, informing which undestroyed volumes are associated with the primary storage and which instance they are allocated to if an exception is thrown during the process.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [X] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

I tried to delete a primary storage with active volumes, then I checked through the logs that all volumes and the VM associated with them were presented.

### CODE REVIEW

**Comment by DaanHoogland** [comment]
not sure if we have any automated way of testing this (yet)

File: server/src/main/java/com/cloud/storage/StorageManagerImpl.java [hunk 1]
```diff
@@ -1200,7 +1200,8 @@ public void doInTransactionWithoutResult(TransactionStatus status) {
                     }
                 });
             } else {
-                throw new CloudRuntimeException("Cannot delete pool " + sPool.getName() + " as there are associated " + "non-destroyed vols for this pool");
+                logger.debug("Cannot delete storage pool {} as the following non-destroyed volumes are on it: {}.", sPool.getName(), getStoragePoolNonDestroyedVolumesLog(sPool.getId()));
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/storage/StorageManagerImpl.java:1203]:
```suggestion
                logger.debug("Cannot delete storage pool {} as the following non-destroyed volumes are on it: {}.", sPool.getName(), () -> getStoragePoolNonDestroyedVolumesLog(sPool.getId()));
```

When executing heavy processes to load data only to log it, call the methods inside a lambda; this way, they will be evaluated only at the right log level.

File: server/src/main/java/com/cloud/storage/StorageManagerImpl.java [hunk 2]
```diff
@@ -1235,17 +1236,16 @@ private boolean deleteDataStoreInternal(StoragePoolVO sPool, boolean forced) {
             if (vlms.first() > 0) {
                 Pair<Long, Long> nonDstrdVlms = volumeDao.getNonDestroyedCountAndTotalByPool(sPool.getId());
                 if (nonDstrdVlms.first() > 0) {
-                    throw new CloudRuntimeException("Cannot delete pool " + sPool.getName() + " as there are associated " + "non-destroyed vols for this pool");
+                    logger.debug("Cannot delete storage pool {} as the following non-destroyed volumes are on it: {}.", sPool.getName(), getStoragePoolNonDestroyedVolumesLog(sPool.getId()));
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/storage/StorageManagerImpl.java:1239]:
```suggestion
                    logger.debug("Cannot delete storage pool {} as the following non-destroyed volumes are on it: {}.", sPool.getName(), () -> getStoragePoolNonDestroyedVolumesLog(sPool.getId()));
```

When executing heavy processes to load data only to log it, call the methods inside a lambda; this way, they will be evaluated only at the right log level.

File: server/src/main/java/com/cloud/storage/StorageManagerImpl.java [hunk 3]
```diff
@@ -1254,7 +1254,8 @@ private boolean deleteDataStoreInternal(StoragePoolVO sPool, boolean forced) {
             // Check if the pool has associated volumes in the volumes table
             // If it does , then you cannot delete the pool
             if (vlms.first() > 0) {
-                throw new CloudRuntimeException("Cannot delete pool " + sPool.getName() + " as there are associated volumes for this pool");
+                logger.debug("Cannot delete storage pool {} as the following non-destroyed volumes are on it: {}.", sPool.getName(), getStoragePoolNonDestroyedVolumesLog(sPool.getId()));
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/storage/StorageManagerImpl.java:1257]:
```suggestion
                logger.debug("Cannot delete storage pool {} as the following non-destroyed volumes are on it: {}.", sPool.getName(), () -> getStoragePoolNonDestroyedVolumesLog(sPool.getId()));
```

When executing heavy processes to load data only to log it, call the methods inside a lambda; this way, they will be evaluated only at the right log level.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Review by BryanMLima — APPROVED** [review]
CLGTM

**Comment by BryanMLima** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@blueorangutan test alma9 kvm-alma9

**Review by SadiJr — APPROVED** [review]
Tested in a local lab. LGTM. In the logs, I can see the volumes and their VMs.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan LLtest

**Comment by DaanHoogland** [comment]
@GutoVeronezi , all satisfied with the current state?

**Comment by DaanHoogland** [comment]
@hsato03 seeing `2024-07-23T18:43:54,232 INFO  [c.c.a.ApiServlet] (qtp253011924-22:[ctx-35b7f332, ctx-24c715d9]) (logid:0c5ca90f) (userId=2 accountId=2 sessionId=node048rydzt
ycoujeeasuf92j07x0) 172.30.205.2 -- GET id=7973fc94-7f84-334d-bf98-8631b37f2b7c&command=deleteStoragePool&response=json 530 Cannot delete pool pr8649-t6961-k
vm-alma8-kvm-pri1 as there are non-destroyed volumes associated to this pool.` means that I don not test the right code, right?
I see none of the new log lines.

**Comment by hsato03** [comment]
> @hsato03 seeing `2024-07-23T18:43:54,232 INFO [c.c.a.ApiServlet] (qtp253011924-22:[ctx-35b7f332, ctx-24c715d9]) (logid:0c5ca90f) (userId=2 accountId=2 sessionId=node048rydzt ycoujeeasuf92j07x0) 172.30.205.2 -- GET id=7973fc94-7f84-334d-bf98-8631b37f2b7c&command=deleteStoragePool&response=json 530 Cannot delete pool pr8649-t6961-k vm-alma8-kvm-pri1 as there are non-destroyed volumes associated to this pool.` means that I don not test the right code, right? I see none of the new log lines.

@DaanHoogland the new logs come right before the exception stack trace:
```
2024-07-23T21:34:23,302 DEBUG [c.c.s.StorageManagerImpl] (qtp1759482496-19:[ctx-84a414ad, ctx-a0b47312]) (logid:921e60ed) Cannot delete storage pool iscsi-1 as the following non-destroyed volumes are on it: [Volume [347f6d86-2816-404d-86bf-d5c21315359f] (attached to VM [808fd88f-06a9-4539-a01f-4c8df1fdcecd]), Volume [f05954b1-8724-4692-bc87-134ca07ba1b0] (attached to VM [da61273a-d56a-4842-aa9d-374e5eac5bbd])].
2024-07-23T21:34:23,305 ERROR [c.c.a.ApiServer] (qtp1759482496-19:[ctx-84a414ad, ctx-a0b47312]) (logid:921e60ed) unhandled exception executing api command: [Ljava.lang.String;@7c2451a8 com.cloud.utils.exception.CloudRuntimeException: Cannot delete pool iscsi-1 as there are non-destroyed volumes associated to this pool.
	at com.cloud.storage.StorageManagerImpl.deleteDataStoreInternal(StorageManagerImpl.java:1495)
...
```

File: server/src/main/java/com/cloud/storage/StorageManagerImpl.java [hunk 4]
```diff
@@ -1514,6 +1515,25 @@ private boolean deleteDataStoreInternal(StoragePoolVO sPool, boolean forced) {
...
+
+        sb.append("[");
+        for (int i = 0; i < nonDestroyedVols.size()-1; i++) {
+            VolumeVO vol = nonDestroyedVols.get(i);
+            volInstance = _vmInstanceDao.findById(vol.getInstanceId());
+            sb.append(String.format("Volume [%s] (attached to VM [%s]), ", vol.getUuid(), volInstance.getUuid()));
+        }
+        VolumeVO lastVol = nonDestroyedVols.get(nonDestroyedVols.size()-1);
+        volInstance = _vmInstanceDao.findById(lastVol.getInstanceId());
+        sb.append(String.format("Volume [%s] (attached to VM [%s])", lastVol.getUuid(), volInstance.getUuid()));
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/storage/StorageManagerImpl.java:1531]:
```suggestion
        for (VolumeVO vol : nonDestroyedVols) {
            volInstance = _vmInstanceDao.findById(vol.getInstanceId());
            sb.append(String.format("Volume [%s] (attached to VM [%s]), ", vol.getUuid(), volInstance.getUuid()));
        }
```

**Review by GutoVeronezi — APPROVED** [review]
LGTM

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan LLtest

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@JoaoJandre at your discretion.

**Pull request merged**